### DISCUSSION TITLE
How to get the Z value of a point on the surface mesh

### DISCUSSION THREAD

**Discussion by eamontoyaa** [pyvista/pyvista#5662] [Q&A] [answered]
Hi there!

I would appreciate a lot any guidance regarding this need:

Let's say I have a mesh surface representing the terrain surface, and I want to know the elevation or $z$ coordinate of a point whose $x$ and $y$ coordinates are known.
It is basically a vertical projection of a (x, y) point on a mesh surface.

Something similar was previously asked [here](https://github.com/pyvista/pyvista-support/issues/174), with no fulfilling answers.

Thanks in advance!

**Comment by banesullivan** [comment, marked as answer]
Here is one way to do it via ray tracing

See https://docs.pyvista.org/version/stable/examples/01-filter/poly-ray-trace#sphx-glr-examples-01-filter-poly-ray-trace-py

```py
import pyvista as pv
from pyvista import examples

mesh = examples.load_random_hills()

# Z bounds of the mesh
bottom, top = mesh.bounds[-2:]
buffer = 1.0

# Define two points that form a line that interesects the mesh
x, y = mesh.center[0:2]
start = [x, y, bottom - buffer]
stop = [x, y, top + buffer]

# Perform ray trace
points, ind = mesh.ray_trace(start, stop)

# Create geometry to represent ray trace
ray = pv.Line(start, stop)
intersection = pv.PolyData(points)

# Render the result
p = pv.Plotter()
p.add_mesh(mesh, show_edges=True, opacity=0.95, color='lightgrey')
p.add_mesh(ray, color="blue", line_width=5)
p.add_mesh(intersection, color="maroon", point_size=10)
p.show()
```

![screenshot](https://github.com/pyvista/pyvista/assets/22067021/37561047-3a5e-483a-8489-978156b9cf12)

**Reply by eamontoyaa** [reply]
``@banesullivan, thanks for your quick reply.

I just tried this nice approach, and your example works as expected. 
Just a note for anybody else who might need this: It does not work for a surface represented by a StructuredGrid, so you have to use `mesh = originally_structured_mesh.extract_geometry()` first:

``` python
import pyvista as pv
from pyvista import examples

# mesh = examples.load_random_hills()
mesh = originally_structured_mesh.extract_geometry()

# Z bounds of the mesh
bottom, top = mesh.bounds[-2:]
buffer = top - bottom

# Define two points that form a line that interesects the mesh
x, y = mesh.center[0:2]
start = [x, y, bottom - buffer]
stop = [x, y, top + buffer]

# Perform ray trace
points, ind = mesh.ray_trace(start, stop)

# Create geometry to represent ray trace
ray = pv.Line(start, stop)
intersection = pv.PolyData(points)

# Render the result
p = pv.Plotter()
p.add_mesh(mesh, show_edges=True, opacity=0.95, color='lightgrey')
p.add_mesh(ray, color="blue", line_width=5)
p.add_mesh(intersection, color="maroon", point_size=10)
p.show()
```

![image](https://github.com/pyvista/pyvista/assets/17820465/ccc6e678-35cb-4fd9-91cb-298c1afda941)
